Tool Interaction Recipes API module

Leo40Git opened the pull request:
This PR creates the Tool Interaction Recipes API module (and also its containing blocks library).

**TODO:**
- [x] Code-based path via registries
- [ ] Data-driven path (requires Resource Loader)

Leo40Git committed:
First draft of in-world recipe API

Leo40Git committed:
Correct identifier namespace

...that's one mouthful of an identifier.

Leo40Git committed:
De-recordify InWorldRecipe to allow for documentation

Apparently you can't add javadoc to record components. What's the point

Leo40Git committed:
add TODO about data-driven recipes

Leo40Git committed:
Transform InWorldRecipe into an interface

tryPerform combines the predicate/consumer duo

Leo40Git committed:
english

TheGlitch76 commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipeRegistries.java:
```diff
@@ -0,0 +1,42 @@
+		return registry.stream()
+				.filter(recipe -> recipe.targetBlock() == targetBlock && recipe.predicate().test(context))
+				.findFirst();
+	}
```
Would this method ever be called in hot code? As it stands right now this is a pretty allocation-heavy and slow way of doing things, but that might be fine.

Leo40Git replied:
I could probably transform this into a regular `for (Entry : registry.getEntries)` ~~(especially since I actually need to change it to return a list).~~

LambdAurora replied:
This definitely needs to be a `for` because the stream API is a disappointment so it will actually go through the entire list and never return early despite you needing just the first match.

BasiqueEvangelist replied:
> I actually need to change it to return a list

Couldn't you add a `doRecipes` method which would try all matching (by block) recipes until one succeeds?

Leo40Git replied:
> Couldn't you add a `doRecipes` method which would try all matching (by block) recipes until one succeeds?

What a coincidence, that's what I just did 🙃

Leo40Git committed:
Simplify code\

Leo40Git committed:
Add mixin method prefixes

FoundationGames commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipe.java:
```diff
@@ -0,0 +1,22 @@
+	/**
+	 * {@return the specific block type required for this recipe}
+	 */
+	@NotNull Block targetBlock();
```
Could this use a `Predicate<BlockState>` rather than just a single Block filter? Some may want to filter by block type (as vanilla recipes do) or filter by a specific block state.

As for the data driven part, the predicate can be specified similar to the way loot tables work
(Example, Revision 2)
```json
{
    "predicates": [
        {
            "type": "quilt:block_state",
            "block": "minecraft:oak_slab",
            "properties": {
                "half": "lower"
            }
        },
        {
            "type": "quilt:block_tag",
            "tag": "minecraft:flowers"
        },
        {
            "type": "custom:custom",
            "custom_condition_data": [ "foo", "bar" ]
        }
    ],
    "result": {
        "block": "minecraft:birch_slab",
        "properties": {}
    }
}
```

Leo40Git replied:
This can be done by checking the `BlockState` in the `tryPerform` method before actually doing anything.

williambl replied:
What if you want to use a block tag as an input?

Leo40Git replied:
...good point. Maybe I should just remove `targetBlock()` entirely.

BasiqueEvangelist replied:
> ...good point. Maybe I should just remove `targetBlock()` entirely.

Couldn't you just make it optional? So that if a recipe will only work for one specific block, that can be optimized, but all other recipes will be run every time?

FoundationGames replied:
> This can be done by checking the `BlockState` in the `tryPerform` method before actually doing anything.

As Will said, now you lose the flexibility of allowing multiple block types via tags (though my json example wouldn't work then)

Leo40Git replied:
> Couldn't you just make it optional? So that if a recipe will only work for one specific block, that can be optimized, but all other recipes will be run every time?

That could work, though I'd need registry events to actually implement that optimization.

williambl replied:
```suggestion
	boolean matches(BlockState state);
```

the most basic impl would just be:
```java
if (this.block != null) {
	return state.getBlock() == this.block;
} else if (this.tag != null) {
	return tag.contains(state.getBlock());
} else return false;
```
but it can be extended to the other blockstate predicate stuff mentioned earlier

BasiqueEvangelist replied:
> the most basic impl would just be:
> 
> ```java
> if (this.block != null) {
> 	return state.getBlock() == this.block;
> } else if (this.tag != null) {
> 	return tag.contains(state.getBlock());
> } else return false;
> ```
> 
> but it can be extended to the other blockstate predicate stuff mentioned earlier

You can just use `tryPerform` for this, instead of adding a new `matches` method.

Leo40Git replied:
Ended up changing `Block targetBlock()` to `Set<Block> targetBlocks()`, I think this resolves the issue nicely (as I said, block state can be checked in `tryPerform`).

Platymemo commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/mixin/PickaxeItemMixin.java:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	@Override
+	public ActionResult useOnBlock(ItemUsageContext context) {
```
overriding methods in mixins is not recommended. I think the ideal path is to inject in a supermixin (mixin to Item) with an empty body and override that handler method in a submixin (this mixin here).

Platymemo commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/mixin/HoeItemMixin.java:
```diff
@@ -0,0 +1,33 @@
+			cancellable = true, locals = LocalCapture.CAPTURE_FAILHARD)
+	public void qsl$doCustomRecipes(ItemUsageContext context, CallbackInfoReturnable<ActionResult> cir,
+									World world, BlockPos blockPos) {
+		if (!InWorldRecipeRegistries.tryPerform(InWorldRecipeRegistries.AXE, context, world.getBlockState(blockPos).getBlock()))
```
```suggestion
		if (!InWorldRecipeRegistries.tryPerform(InWorldRecipeRegistries.HOE, context, world.getBlockState(blockPos).getBlock()))
```

Leo40Git replied:
oops, already fixed this locally! copy+paste is a double-edged sword

Platymemo commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/mixin/ShovelItemMixin.java:
```diff
@@ -0,0 +1,33 @@
+			cancellable = true, locals = LocalCapture.CAPTURE_FAILHARD)
+	public void qsl$doCustomRecipes(ItemUsageContext context, CallbackInfoReturnable<ActionResult> cir,
+									World world, BlockPos blockPos, BlockState blockState, PlayerEntity playerEntity) {
+		if (!InWorldRecipeRegistries.tryPerform(InWorldRecipeRegistries.AXE, context, blockState.getBlock()))
```
```suggestion
		if (!InWorldRecipeRegistries.tryPerform(InWorldRecipeRegistries.SHOVEL, context, blockState.getBlock()))
```

Platymemo commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/mixin/SwordItemMixin.java:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	@Override
+	public ActionResult useOnBlock(ItemUsageContext context) {
```
Same comment as with the pickaxe mixin, override is not recommended.

Platymemo reviewed (commented):
How do we feel about including damage effects as default in `tryPerform` so data recipes can customize how much damage they inflict on the items used? Will need javadocs to recommend calling super

Leo40Git committed:
Use multimap-based approach

Allow multiple target blocks

Leo40Git committed:
Override supermixin handler instead of method directly

Leo40Git committed:
okay, override a method _called_ by the injection, not the injection itself

Leo40Git committed:
oops, forgot to update javadoc

Leo40Git committed:
Add SimpleInWorldRecipe

BasiqueEvangelist commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipe.java:
```diff
@@ -4,19 +4,26 @@
 	 * @return {@code true} if recipe was successfully performed, {@code false} otherwise
 	 */
 	boolean tryPerform(@NotNull ItemUsageContext context);
+
+	/**
+	 * Invoked on data pack reload. Use this to update your {@linkplain #targetBlocks() target block set}.
+	 */
+	default void onReload() { }
```
Is this needed? You could just recalculate the set every time in `targetBlocks`, considering it'll only be used on datapack reload to recreate the multimap.

Leo40Git replied:
Fair point, method yeeted.

Leo40Git committed:
Remove InWorldRecipe.onReload

Getting a tag's content is pretty lightweight, all things considered

Leo40Git committed:
De-hardcode "damage item by 1" logic

It's now the default implementation of InWorldRecipe.tryPerform
Encourage calling super on success

Platymemo commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/mixin/ItemMixin.java:
```diff
@@ -0,0 +1,22 @@
+public abstract class ItemMixin {
+	@Inject(method = "useOnBlock", at = @At("HEAD"), cancellable = true)
+	public void qsl$useOnBlockOverride(ItemUsageContext context, CallbackInfoReturnable<ActionResult> cir) {
+		cir.setReturnValue(qsl$useOnBlock(context));
```
This affects every item's useOnBlock method that isn't overridden. You want this to not change vanilla behavior unnecessarily.
```suggestion
		ActionResult result = qsl$useOnBlock(context);
		if (result.isAccepted()) {
				cir.setReturnValue(result);
		}
```

Platymemo replied:
Also, overriding the handler method was my original intention, but this also works just as well

Leo40Git replied:
It returns `ActionResult.PASS` by default, which is the same behavior as vanilla (and I doubt this will change, unless Mojang wants to break interacting with _every single block_).

Leo40Git replied:
I'll allow it to return `null`, which will cause it to defer to the actual method.

Platymemo replied:
oops, for some reason I thought it returned `ActionResult.FAIL`. My b

Platymemo replied:
I think `ActionResult.PASS` should defer to the actual method tho, and only `ActionResult.SUCCESS` or `ActionResult.FAIL` should cause it to deviate

Leo40Git commented:
Adding a "stupid simple" recipe now looks like:
```java
InWorldRecipeRegistries.PICKAXE.register(new Identifier("coolmod", "pick_cool_ore"), new SimpleInWorldRecipe(CoolModBlocks.COOL_ORE) {
    @Override
    public boolean tryPerform(@NotNull ItemUsageContext context) {
        World world = context.getWorld();
        world.setBlockState(context.getBlockPos(), Blocks.STONE.getDefaultState(), 3);
        return super.tryPerform(context);
    }
});
```

Leo40Git committed:
Defer to actual Item.useOnBlock if PASS is returned

Leo40Git committed:
Prepare for data-driven implementation

i509VCB commented:
Could you explain the use case for this module?

Also given we are still quite early and need core things in place in QSL your PR may not be immediately addressed

i509VCB commented:
Oops

Leo40Git commented:
> Could you explain the use case for this module?

Normally, adding a stripping/tilling/shoveling(?) recipe requires a mixin. This module reduces the number of mixins from `number_of_mods_that_add_those_kinds_of_recipes` to one.

Also, this extends stripping and shoveling to support more than just "replace any state of this block with this state of this other block" (or even "with the default state of this block" in the case of stripping, though that's because it does property copying by default), and it adds "right click recipe actions" to pickaxes and swords as well.

i509VCB commented:
So this is really a module to act as a registry for recipe actions based on tool actions?

Leo40Git commented:
Yes. I was asking about a better name in the Discord, "tool interaction recipe" was suggested.

i509VCB commented:
Okay, since this involves recipes, @shedaniel do you have any particular requirements for REI et al. for custom recipe systems?

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipeRegistries.java:
```diff
@@ -0,0 +1,31 @@
+public final class InWorldRecipeRegistries {
+	private InWorldRecipeRegistries() { throw new AssertionError(); }
+
+	public static final Registry<InWorldRecipe> SWORD = create("sword");
```
These seem to be based on tool and not action, such as tilling, pathing and debarking. Is there a reason for this name scheme? Or is it more abstract than those specific actions.

Leo40Git replied:
It's more abstract than those specific actions, I feel.

Like, right clicking with a sword could mean carving, stabbing, slicing, etc... the same ambiguity could be applied to the other tool types.

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipeRegistries.java:
```diff
@@ -0,0 +1,31 @@
+		RegistryKey<Registry<InWorldRecipe>> key =
+				RegistryKey.ofRegistry(new Identifier("qsl-blocks-in-world-recipes", "in_world/" + toolName));
+		Registry<InWorldRecipe> registry = new SimpleRegistry<>(key, Lifecycle.stable());
+		// FIXME add to root registry - not working for some reason???
```
This is because the root registry stuff is a little more complex than that. Soon to come registry api will handle that

Leo40Git replied:
This comment was because `Registry.register(Registry.REGISTIRES, Identifier, Registry<InWorldRecipe>)` wasn't resolving.

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/impl/InWorldRecipeMaps.java:
```diff
@@ -0,0 +1,50 @@
+
+	// TODO update this to run on datapack reload
+	// TODO data-drive!
+	public static void update() {
```
Could this cause a thread safety issue if the client uses these maps and tries to display the recipe to a player and the server signals an update, especially on integrated server.

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/impl/InWorldRecipeMaps.java:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	private static Multimap<Block, InWorldRecipe> fromRegistry(@NotNull Registry<InWorldRecipe> registry) {
+		//noinspection UnstableApiUsage
```
A note: these noinspection lines are not handled by eclipse and the java language server fyi.

Leo40Git replied:
Will convert to an annotation.

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/BlockPredicateUtil.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public final class BlockPredicateUtil {
```
Hmm what is the primary purpose of the util method in here?

Leo40Git replied:
This is intended for the future data-driven implementation, so we can extract the block set from a recipe's predicate(s) to put them in the multimap.

If a recipe has no block predicates (or the `ANY` block predicate), it'll be added to an alternate list (NYI).

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/BlockPredicateUtil.java:
```diff
@@ -0,0 +1,21 @@
+	public @NotNull Set<Block> getBlocks(@NotNull BlockPredicate predicate) {
+		Tag<Block> tag = ((BlockPredicateAccessor) predicate).getTag();
+		if (tag != null)
+			return new HashSet<>(tag.values()); // /shrug
```
Just do oneliner `if (condition) statement;` if you aren't going to put `{ .. }` on these statements

i509VCB commented on library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/impl/InWorldRecipeMaps.java:
```diff
@@ -0,0 +1,50 @@
+package org.quiltmc.qsl.inworldrecipes.impl;
+
+import com.google.common.collect.*;
```
Please avoid star imports

Leo40Git replied:
damn you auto-import

i509VCB reviewed (commented):
Not done yet, trying to address api surface first

shedaniel commented:
I am not too sure about the reasoning and scope of this API, the implementation is also, weird.

1. InWorldRecipe, not sure what purpose does targetBlocks serve, it may be for faster lookup, but could be extremely limiting. Even worse is that in InWorldRecipeMaps#tryPerform, it is doing a secondary contains check, I know it would be O(1), but this seems quite redundant and the direction of the api is confusing.

2. InWorldRecipe could not be easily extended by mods (this is a simple fix), InWorldRecipe#tryPerform offers no easy way for mods to alter the functionality. You could possibly make it so there would be an abstract method to do extra things, or just leave the default implementation out.

3. Types of items, this seems to be quite limiting, with each type of tool holding a registry and map, and a mixin for every type of tool. I am not sure but I personally think that it would even be easier for modders to listen to an item use event and have their own system

4. I am not sure how you would data driven this without being more complicated then just allowing people to extend the vanilla recipes / serializers system, vanilla recipes syncs to the clients for you, etc

As for REI requirements, any mods that wish to add REI functionality for something like this would need to have their own plugin for showing the actual functionality (placing widgets, rendering texts, etc), mods would just need to have this info on the client.
(so syncing data driven content to client)

Also, I am not sure why you need a registry and then copy to a map

Leo40Git commented:
> InWorldRecipe, not sure what purpose does targetBlocks serve, it may be for faster lookup, but could be extremely limiting.

I'll implement an alternate iteration-based path for recipes that doesn't require a set of target blocks.

> Even worse is that in InWorldRecipeMaps#tryPerform, it is doing a secondary contains check,

That's a leftover from an earlier version, oops

> InWorldRecipe#tryPerform offers no easy way for mods to alter the functionality.

??? It can be overriden entirely without calling super (if you don't want the tool to get damaged or want it to be damaged by like 10 or something), not sure what this means - it's entirely customizable, all it needs to do is to return `true` to indicate success or `false` to indicate failure (and failure means the game will try to apply another recipe).

> I am not sure but I personally think that it would even be easier for modders to listen to an item use event and have their own system

I'll probably make the system use an item use event rather than this "mix into each and every tool class" approach.

> I am not sure how you would data driven this

It'll be a fun experience to learn how to do this.

Leo40Git commented:
Renaming to "tool interaction recipes".

Leo40Git committed:
Rename to "tool interaction recipes"

Leo40Git committed:
Add ToolInteractionRecipeCollection

A more convenient way to store and invokes tool interaction recipes
Also implements the iteration-based approach for recipes which allow all block types

Prepare for an item use event style approach
(TODO: figure out how to check tool type and use correct recipe collection based on that)

Leo40Git committed:
add note about replacing mixin

Leo40Git committed:
follow style

Leo40Git committed:
Implement new approach

Leo40Git commented:
~~Screw it I'm adding a block render layer thing as well~~

nvm it just muddles up this PR needlessly

Leo40Git committed:
Add Block Render Layer API

Leo40Git commented on library/blocks/render-layers/src/main/java/org/quiltmc/qsl/blockrenderlayer/mixin/RenderLayersMixin.java:
```diff
@@ -0,0 +1,22 @@
+	@Shadow @Final private static Map<Block, RenderLayer> BLOCKS;
+	@Shadow @Final private static Map<Fluid, RenderLayer> FLUIDS;
+
+	static {
```
I'm assuming this is executed _after_ the pre-existing static block? Not sure how static blocks work in mixin, honestly.

BasiqueEvangelist commented:
Why does this have the two modules together? Wouldn't it be better to PR render layers separately, considering that it is ported from FAPI, while in-world recipes is a new module?

Leo40Git committed:
Revert "Add Block Render Layer API"

This reverts commit c4e83a87f007f26e574330721655b031ce787801.

Leo40Git commented:
Closing for now, will be reopened/recreated later.
